addons/sale_timesheet/models/sale_order.py

```diff
@@ -72,6 +73,32 @@ def _compute_field_value(self, field):
                 # We want to display only one time the warning for each SOL
                 upsellable_lines.write({'has_displayed_warning_upsell': True})
 
+    def _compute_show_hours_recorded_button(self):
+        if not self.env.user.has_group('hr_timesheet.group_hr_timesheet_user'):
+            for order in self:
+                order.show_hours_recorded_button = False
+            return
+
+        orders_to_check = self.filtered(lambda l: l.state not in ['draft', 'sent'])
```

Comment:
Could you check if something like this could work?
```py
sols_read_group = self.env['sale.order.line']._read_group([
    ('order_id', 'in', self.ids),
    ('order_id.state', 'not in', ['draft', 'sent']),
    ('product_id.type', '=', 'service'),
    ('product_id.service_type', 'not in', ['milestones', 'manual']),
    ('product_id.invoice_policy', '!=', 'delivery'),
], ['order_id'], ['order_id'])
show_button_ids = {sols['order_id'][0] for sols in sols_read_group}
for order in self:
    order.show_hours_recorded_button = order.project_count and order.id in show_button_ids
```
(I didn't test this code)